clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3187,3 +3195,68 @@ LangAS CIRGenModule::getGlobalVarAddressSpace(const VarDecl *D) {
 
   return getTargetCIRGenInfo().getGlobalVarAddressSpace(*this, D);
 }
+
+mlir::ArrayAttr CIRGenModule::buildAnnotationArgs(AnnotateAttr *attr) {
+  ArrayRef<Expr *> exprs = {attr->args_begin(), attr->args_size()};
+  if (exprs.empty()) {
+    return mlir::ArrayAttr::get(builder.getContext(), {});
+  }
+  llvm::FoldingSetNodeID id;
+  for (Expr *e : exprs) {
+    id.Add(cast<clang::ConstantExpr>(e)->getAPValueResult());
```

Comment:
While debugging some core dump in this function, I wonder what happens here when e is not a `clang::ConstantExpr`?
@ghehg an idea about the missing comments here? :smile:
No need for some `IgnoreParenCasts()` here?

Reply:
I am still diving into it. Please ignore my comments since I got the inheritance wrong.